### PR TITLE
Include option to train models on taskcluster

### LINKED ISSUE #3688: Train models on TaskCluster whenever their corresponding files are modified
Whenever a file in the `bugbug/models` directory is modified, we could have a task to train the corresponding model.

For example, if a new Pull Request is modifying the `bugbug/models/defect.py` file, the task should train the `defect` model:
```
python -m scripts.trainer defect
```

### PR DESCRIPTION
**Pull request by jpangas**
**Issue:**
Closes #3688 

**Comments:**
This PR is a work in progress and hence a draft PR. I wanted to get some feedback before proceeding just to ensure I am on the right track.

### CODE REVIEW

**Comment by jpangas** [comment]
@marco-c , @suhaibmujahid , I wanted to get some feedback before proceeding just to ensure I am on the right track.

File: scripts/trainer.py [hunk 1]
```diff
@@ -153,8 +198,32 @@ def parse_args(args):
...
+                    "-m",
+                    "scripts.trainer",
+                    args.model,
+                    "--limit",
+                    str(args.limit),
+                    "--no-download" if not args.download_db else "",
+                    "--download-eval" if args.download_eval else "",
+                    "--lemmatization" if args.lemmatization else "",
+                    "--classifier",
+                    args.classifier,
```
Comment (suhaibmujahid) [inline scripts/trainer.py:215]: I would avoid hard-coping the args list here. We could reuse the original command, just stripping `--taskcluster` from it.

File: scripts/trainer.py [hunk 2]
```diff
@@ -16,6 +18,44 @@
...
+    task_id = taskcluster.utils.slugId()
+
+    try:
+        queue.createTask(task_id, task_definition)
+        logger.info(
+            "Task created at https://community-tc.services.mozilla.com/tasks/%s",
+            task_id,
+        )
+    except taskcluster.exceptions.TaskclusterAuthFailure:
+        logger.exception("Failed to authenticate with Taskcluster")
```
Comment (suhaibmujahid) [inline scripts/trainer.py:56]: How do we handle the auth?

**Review by suhaibmujahid — COMMENTED** [review]
Thank you, @jpangas!

The main missing part is pulling the changes for the remote repository. Also, I did not use `taskcluster` API before, could you test if the API usage is correct?

**Comment by jpangas** [comment]
> Thank you, @jpangas!
> 
> The main missing part is pulling the changes for the remote repository. Also, I did not use `taskcluster` API before, could you test if the API usage is correct?

@suhaibmujahid, I have tested it locally, currently facing this exception `taskcluster.exceptions.TaskclusterRestFailure: This request requires Taskcluster credentials that satisfy the following scope expression:`

File: scripts/trainer.py (hunk 1 above)
Comment (jpangas) [reply scripts/trainer.py]: I have done this. Please check the latest `command` and let me know.

**Comment by jpangas** [comment]
I need to know which are the right scopes to set in this task. @marco-c

**Comment by suhaibmujahid** [comment]
> I have tested it locally, currently facing this exception `taskcluster.exceptions.TaskclusterRestFailure: This request requires Taskcluster credentials that satisfy the following scope expression:`

Could you share the full error message?

> I need to know which are the right scopes to set in this task.

Is this related to the above message?

File: scripts/trainer.py (hunk 1 above)
Comment (suhaibmujahid) [reply scripts/trainer.py]: Thank you!

File: scripts/trainer.py (hunk 2 above)
Comment (suhaibmujahid) [reply scripts/trainer.py]: > Have I done it correctly?

I do not know 😅 I did not use the API before. From the docs, it seems that the auth need to be done differentl: https://github.com/taskcluster/taskcluster/tree/main/clients/client-py#authentication-options

**Comment by jpangas** [comment]
Yes, it is related to the message above. 
Here is the full error message:
<details>
<summary>Error Message </summary>
<br>

```
taskcluster.exceptions.TaskclusterRestFailure: This request requires Taskcluster credentials that satisfy the following scope expression:

{
  "AllOf": [
    "queue:create-task:project:none",
    "queue:scheduler-id:-",
    {
      "AnyOf": [
        "queue:create-task:highest:proj-bugbug/compute-smaller",
        "queue:create-task:very-high:proj-bugbug/compute-smaller",
        "queue:create-task:high:proj-bugbug/compute-smaller",
        "queue:create-task:medium:proj-bugbug/compute-smaller",
        "queue:create-task:low:proj-bugbug/compute-smaller",
        "queue:create-task:very-low:proj-bugbug/compute-smaller",
        "queue:create-task:lowest:proj-bugbug/compute-smaller"
      ]
    }
  ]
}


---

* method:     createTask
* errorCode:  InsufficientScopes
* statusCode: 403
* time:       2023-10-13T07:04:28.274Z
(bugbug) (base) john@swiftyxswaggy:~/Documents/Local_Code/mozilla/bugbug$ 
```
</details>

File: scripts/trainer.py (hunk 2 above)
Comment (jpangas) [reply scripts/trainer.py]: You are right! I checked the `auth` credentials and they weren't present.  I will need to create them, I read in the documentation that-the authentication information passed with each API call is used to determine a set of scopes associated with the caller.
It's my first time using the API too. 😅 I am a little confused on the right credentials to create.

**Comment by marco-c** [comment]
Here's an example of a standalone script doing Taskcluster authentication:
https://github.com/mozilla/code-review/blob/f0aedde5e3f5d5ac196310ddbacdbf5e5922cd8a/bot/tools/fix_missing.py#L12
https://github.com/mozilla/code-review/blob/f0aedde5e3f5d5ac196310ddbacdbf5e5922cd8a/bot/tools/fix_missing.py#L26 (you'll have to use the community-tc URL here instead of firefox-ci)
https://github.com/mozilla/code-review/blob/f0aedde5e3f5d5ac196310ddbacdbf5e5922cd8a/bot/tools/fix_missing.py#L160

**Pull request closed without merging**